From 1f2fd2598b8f11e6f52c8c71ce11fb208cb7e7d7 Mon Sep 17 00:00:00 2001
From: Hoss <hoss@hardcoretech.co>
Date: Fri, 14 Aug 2020 16:38:33 +0800
Subject: [PATCH] Support Django > 3

1. Django builtin six library was removed since v3.0.0
2. FieldDoesNotExist was moved to django.core.exceptions since v1.8.0, and its comp module was removed since v3.1.0
---
 AUTHORS.rst                   | 1 +
 django_pandas/io.py           | 9 ++++++++-
 django_pandas/tests/models.py | 2 +-
 setup.py                      | 1 +
 4 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/AUTHORS.rst b/AUTHORS.rst
index fe2fcf0..67a1804 100644
--- a/AUTHORS.rst
+++ b/AUTHORS.rst
@@ -27,3 +27,4 @@ Contributions
 - `Anthony Monthe <https://github.com/ZuluPro>`_
 - `Vincent Toupet <https://github.com/vtoupet>`_
 - `Anton Ian Sipos <https://github.com/aisipos>`_
+- `Chuan-Jhe Hwong <https://github.com/CJHwong>`_
diff --git a/django_pandas/io.py b/django_pandas/io.py
index 5225977..e451dd9 100644
--- a/django_pandas/io.py
+++ b/django_pandas/io.py
@@ -3,13 +3,20 @@
 import django
 
 
+FieldDoesNotExist = (
+    django.db.models.fields.FieldDoesNotExist
+    if django.VERSION < (1, 8)
+    else django.core.exceptions.FieldDoesNotExist
+)
+
+
 def to_fields(qs, fieldnames):
     for fieldname in fieldnames:
         model = qs.model
         for fieldname_part in fieldname.split('__'):
             try:
                 field = model._meta.get_field(fieldname_part)
-            except django.db.models.fields.FieldDoesNotExist:
+            except FieldDoesNotExist:
                 try:
                     rels = model._meta.get_all_related_objects_with_model()
                 except AttributeError:
diff --git a/django_pandas/tests/models.py b/django_pandas/tests/models.py
index 0dca23d..ce57ece 100644
--- a/django_pandas/tests/models.py
+++ b/django_pandas/tests/models.py
@@ -1,5 +1,5 @@
 from django.db import models
-from django.utils.encoding import python_2_unicode_compatible
+from six import python_2_unicode_compatible
 from django_pandas.managers import DataFrameManager, PassThroughManager
 
 
diff --git a/setup.py b/setup.py
index 8e7ce97..456f222 100644
--- a/setup.py
+++ b/setup.py
@@ -23,6 +23,7 @@
     packages=find_packages(),
     install_requires=[
         'pandas>=0.14.1',
+        'six>=1.15.0',
     ],
     classifiers=[
         'Development Status :: 3 - Alpha',
